File: src/store/local.ts
import type { StorageBackend, MemoryEntry, RetrievalQuery } from '../core/types.js';

/**
 * In-memory storage backend. Zero dependencies.
 * Perfect for development, testing, and single-process agents.
 * Data is lost when the process exits.
 */
export class LocalStore implements StorageBackend {
  private entries: Map<string, MemoryEntry> = new Map();

  async get(id: string): Promise<MemoryEntry | null> {
    return this.entries.get(id) ?? null;
  }

  async getAll(scope?: string): Promise<MemoryEntry[]> {
    const all = Array.from(this.entries.values());
    if (!scope) return all;
    return all.filter(e => e.scope === scope);
  }

  async set(entry: MemoryEntry): Promise<void> {
    this.entries.set(entry.id, { ...entry });
  }

  async delete(id: string): Promise<boolean> {
    return this.entries.delete(id);
  }

  async clear(scope?: string): Promise<void> {
    if (!scope) {
      this.entries.clear();
      return;
    }
    for (const [id, entry] of this.entries) {
      if (entry.scope === scope) {
        this.entries.delete(id);
      }
    }
  }

  async search(query: RetrievalQuery): Promise<MemoryEntry[]> {
    let results = Array.from(this.entries.values());

    if (query.scope) {
      results = results.filter(e => e.scope === query.scope);
    }

    if (query.types && query.types.length > 0) {
      results = results.filter(e => query.types!.includes(e.type));
    }

    return results;
  }

  /** Get the number of stored entries */
  get size(): number {
    return this.entries.size;
  }
}

File: python/agentic_memory/store/local.py
"""In-memory storage backend. Zero dependencies."""

from __future__ import annotations

from typing import Optional

from ..types import MemoryEntry, RetrievalQuery


class LocalStore:
    """
    In-memory storage backend.
    Perfect for development, testing, and single-process agents.
    Data is lost when the process exits.
    """

    def __init__(self) -> None:
        self._entries: dict[str, MemoryEntry] = {}

    async def get(self, id: str) -> Optional[MemoryEntry]:
        entry = self._entries.get(id)
        return entry if entry is None else MemoryEntry(**vars(entry))

    async def get_all(self, scope: Optional[str] = None) -> list[MemoryEntry]:
        entries = list(self._entries.values())
        if scope is not None:
            entries = [e for e in entries if e.scope == scope]
        return entries

    async def set(self, entry: MemoryEntry) -> None:
        self._entries[entry.id] = MemoryEntry(**vars(entry))

    async def delete(self, id: str) -> bool:
        if id in self._entries:
            del self._entries[id]
            return True
        return False

    async def clear(self, scope: Optional[str] = None) -> None:
        if scope is None:
            self._entries.clear()
        else:
            to_delete = [k for k, v in self._entries.items() if v.scope == scope]
            for k in to_delete:
                del self._entries[k]

    async def search(self, query: RetrievalQuery) -> list[MemoryEntry]:
        results = list(self._entries.values())
        if query.scope is not None:
            results = [e for e in results if e.scope == query.scope]
        if query.types:
            results = [e for e in results if e.type in query.types]
        return results

    @property
    def size(self) -> int:
        return len(self._entries)
